torch/distributed/_composable/fsdp/_fsdp_collectives.py

```diff
@@ -215,6 +215,10 @@ def foreach_reduce(
                 fsdp_param.sharded_param.grad += new_sharded_dtensor_grad
             else:
                 fsdp_param.sharded_param.grad = new_sharded_dtensor_grad
+            for hook in getattr(
```

Comment:
We can see the lines above are FSDP2's manual gradient accumulation. This added logic below tries to run any registered hooks if they exist.

The user will see `fsdp_param.sharded_param`, which is the sharded `DTensor` with its accompanying sharded `DTensor` gradient.